test/threadsafe_function/threadsafe_function_existing_tsfn.cc

```diff
@@ -0,0 +1,107 @@
+#include "napi.h"
+#include <cstdlib>
+
+#if (NAPI_VERSION > 3)
+
+using namespace Napi;
+
+namespace {
+
+struct TestContext {
+  TestContext(Promise::Deferred &&deferred) : deferred(std::move(deferred)){};
+  napi_threadsafe_function tsfn;
+  Promise::Deferred deferred;
+};
+
+void FinalizeCB(napi_env /*env*/, void *context, void * /*finalizedata*/) {
+  TestContext *testContext = static_cast<TestContext *>(context);
+  delete testContext;
+}
+
+void CallJSWithData(napi_env env, napi_value value, void *context, void *data) {
+  TestContext *testContext = static_cast<TestContext *>(context);
+  double *dblVal = static_cast<double *>(data);
+  Function(env, value).Call({Number::New(env, *dblVal)});
+  delete dblVal;
+
+  napi_status status =
+      napi_release_threadsafe_function(testContext->tsfn, napi_tsfn_release);
+
+  NAPI_THROW_IF_FAILED_VOID(env, status);
+
+  testContext->deferred.Resolve(Env(env).Undefined());
+}
+
+void CallJSNoData(napi_env env, napi_value value, void *context,
+                  void * /*data*/) {
+  TestContext *testContext = static_cast<TestContext *>(context);
+  Value undef = Env(env).Undefined();
+  Function(env, value).Call({undef});
```

Comment:
If we're using `Napi::Promise`, is the callback necessary here? Or we could promisify the function signature in the JS, by which the test could be more straightforward to be read. 

i.e. using one single future style might reduce unrelated synchronization code.

Reply:
The `Napi::Promise` is used to synchronize the finalization of the tsfn (resolved in tsfn Finalizer), whereas the callback `Function` is executed in order to update that `lastResult`, which is checked after tsfn finalization (ie. `await testCall()`)

If we did not use the callback, we would need the `Napi::Promise` to resolve with this correct value (either `undefined` or `std::rand()`) based off the call used, but still wait on the TSFN destruction (ie still in the finalizer). I think this would add additional complexity...?

Do you have suggestions?

Reply:
Addressed.  `CallJSWithData` / `CallJSNoData` sets a value stored in the test's context. The tsfn finalizer resolves the `Napi::Promise`.